injector/metapodtemplate.go

```diff
@@ -0,0 +1,266 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/servicebinding/service-binding-controller/apis/v1alpha3"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/jsonpath"
+)
+
+// MetaPodTemplate contains the subset of a PodTemplateSpec that is appropriate for service binding.
+type MetaPodTemplate struct {
+	workload runtime.Object
+	mapping  *v1alpha3.ClusterWorkloadResourceMappingTemplate
+
+	Annotations map[string]string
+	Containers  []MetaContainer
+	Volumes     []corev1.Volume
+}
+
+// MetaContainer contains the aspects of a Container that are appropriate for service binding.
+type MetaContainer struct {
+	Name         string
+	Env          []corev1.EnvVar
+	VolumeMounts []corev1.VolumeMount
+}
+
+// NewMetaPodTemplate coerces the workload object into a MetaPodTemplate following the mapping definition. The
+// resulting MetaPodTemplate may have one or more service bindings applied to it at a time, but should not be reused.
+// The workload must be JSON marshalable.
+func NewMetaPodTemplate(ctx context.Context, workload runtime.Object, mapping *v1alpha3.ClusterWorkloadResourceMappingTemplate) (*MetaPodTemplate, error) {
+	mpt := &MetaPodTemplate{
+		workload: workload,
+		mapping:  mapping,
+
+		Annotations: map[string]string{},
+		Containers:  []MetaContainer{},
+		Volumes:     []corev1.Volume{},
+	}
+
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(workload)
+	if err != nil {
+		return nil, err
+	}
+	uv := reflect.ValueOf(u)
+
+	if err := mpt.getAt(mpt.mapping.Annotations, uv, &mpt.Annotations); err != nil {
+		return nil, err
+	}
+	for i := range mpt.mapping.Containers {
+		cp := jsonpath.New("")
+		if err := cp.Parse(fmt.Sprintf("{%s}", mpt.mapping.Containers[i].Path)); err != nil {
+			return nil, err
+		}
+		cr, err := cp.FindResults(u)
+		if err != nil {
+			// errors are expected if a path is not found
+			continue
+		}
+		for _, cv := range cr[0] {
+			mc := MetaContainer{
+				Name:         "",
+				Env:          []corev1.EnvVar{},
+				VolumeMounts: []corev1.VolumeMount{},
+			}
+
+			if mpt.mapping.Containers[i].Name != "" {
+				// name is optional
+				if err := mpt.getAt(mpt.mapping.Containers[i].Name, cv, &mc.Name); err != nil {
+					return nil, err
+				}
+			}
+			if err := mpt.getAt(mpt.mapping.Containers[i].Env, cv, &mc.Env); err != nil {
+				return nil, err
+			}
+			if err := mpt.getAt(mpt.mapping.Containers[i].VolumeMounts, cv, &mc.VolumeMounts); err != nil {
+				return nil, err
+			}
+
+			mpt.Containers = append(mpt.Containers, mc)
+		}
+	}
+	if err := mpt.getAt(mpt.mapping.Volumes, uv, &mpt.Volumes); err != nil {
+		return nil, err
+	}
+
+	return mpt, nil
+}
+
+// WriteToWorkload applies mutation defined on the MetaPodTemplate since it was created to the workload resource the
+// MetaPodTemplate was created from. This method should generally be called once per instance.
+func (mpt *MetaPodTemplate) WriteToWorkload(ctx context.Context) error {
```

Comment:
`ctx` is not used anywhere.

Reply:
I tend to add a context to exported methods that might possibly use it in the future. For example, if we ever need to make a remote call or lookup a logger. I've been burned too many times by breakages when the author realizes too late that they actually needed a context.